### PR TITLE
[Improvement-15910][UI] Supposed to provide a default value for the custom parallelism when using the mode of parallel execution.

### LINKED ISSUE #15910: [Improvement][UI] Supposed to provide a default value for the custom parallelism when using the mode of parallel execution.
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

Currently the system hasn't provided a default value for the custom parallelism yet. It's really risky If the user chooses a long period to run the task by the mode of parallel execution. So it's required to provide a default value and max value for the custom parallelism.
image

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by calvinjiang**
## Purpose of the pull request
This PR will close #15910 .

## Verify this pull request
This change added tests and can be verified as follows:
- *Manually verified the change by testing locally.* 

The test result is as follows:

image

### CODE REVIEW

File: dolphinscheduler-ui/src/views/projects/workflow/definition/components/start-modal.tsx
Comment (ruanwenjun) [inline dolphinscheduler-ui/src/views/projects/workflow/definition/components/start-modal.tsx]: It's not a good idea to restrict the parallelism should between `1` and `100`, this shouldn't be restricted.

File: dolphinscheduler-ui/src/views/projects/workflow/definition/components/use-form.ts
Comment (ruanwenjun) [inline dolphinscheduler-ui/src/views/projects/workflow/definition/components/use-form.ts:69]: Why the default value is 2 rather than 3? If you want to set a default value, I would prefer 1 or a default empty value.

**Comment by ruanwenjun** [comment]
Please don't add `ready-to-merge` label before receive enough committers' approve.

File: dolphinscheduler-ui/src/views/projects/workflow/definition/components/start-modal.tsx
Comment (zhuangchong) [reply dolphinscheduler-ui/src/views/projects/workflow/definition/components/start-modal.tsx]: I understand that an upper limit should be set. If there is no upper limit, all complement tasks will be triggered, which will put a lot of pressure on the system.

File: dolphinscheduler-ui/src/views/projects/workflow/definition/components/use-form.ts
Comment (zhuangchong) [reply dolphinscheduler-ui/src/views/projects/workflow/definition/components/use-form.ts:69]: I understand that if parallel mode is selected, it should be greater than 1, otherwise just select serial mode. Having default values ​​will make the user experience more friendly and allow users to do as few operations as possible.

File: dolphinscheduler-ui/src/views/projects/workflow/definition/components/start-modal.tsx
Comment (zhuangchong) [reply dolphinscheduler-ui/src/views/projects/workflow/definition/components/start-modal.tsx]: I have encountered a real scenario where the user needs to supplement three years of data. After selecting the parallel mode, 1500+ workflow instances are generated at one time, which directly causes the dolphinscheduler worker service to go down. I suggest that there is an upper limit to limit it. We can discuss this upper limit. How much is appropriate to set.

Comment (ruanwenjun) [reply dolphinscheduler-ui/src/views/projects/workflow/definition/components/start-modal.tsx]: Master/Worker has its own protection strategy, I don't think we should rely on a workflow config to protect the system. And I don't think we can get the correct upper limit at this discussion, different user have different cluster size and usage Scenarios, if we don't know how to set the max value, why we need to set. In additional, if we want to avoid the system crash in a large parallel, the correct way is to limit the rate at server side.

File: dolphinscheduler-ui/src/views/projects/workflow/definition/components/use-form.ts
Comment (ruanwenjun) [reply dolphinscheduler-ui/src/views/projects/workflow/definition/components/use-form.ts:69]: If my habit is to use maximum parallelism every time, then this change will not be friendly to me.

Comment (calvinjiang) [reply dolphinscheduler-ui/src/views/projects/workflow/definition/components/use-form.ts:69]: > If my habit is to use maximum parallelism every time, then this change will not be friendly to me.

You could also use the maximum parallelism 100. But limit-free will be risky to the system.

File: dolphinscheduler-ui/src/views/projects/workflow/definition/components/start-modal.tsx
Comment (Gallardot) [reply dolphinscheduler-ui/src/views/projects/workflow/definition/components/start-modal.tsx]: If an appropriate value cannot be negotiated, it should be returned by `api-server` and allow the administrator to configure it.

File: dolphinscheduler-ui/src/views/projects/workflow/definition/components/use-form.ts
Comment (ruanwenjun) [reply dolphinscheduler-ui/src/views/projects/workflow/definition/components/use-form.ts:69]: If this will bring risky, there will exist a lot of risky in the system. We need to limit the scheduler number, limit the task number in one workflow, limit the size (parallelism in each backfill * backfill request number)...... The main thing is that why this will bring risky to the system, if this will bring risky to system we should fix this in a correct way, rather than simple limit the input form, can this restriction fix this problem? From my point, this restriction just make it more inconvinent for me to use, but it does not solve the problem caused by the high stress.

Comment (zhuangchong) [reply dolphinscheduler-ui/src/views/projects/workflow/definition/components/use-form.ts:69]: Not using the maximum parallelism has another meaning, ensuring the execution of other normal scheduled tasks. If the user directly uses the maximum degree of parallelism, the normally scheduled tasks will not be executed normally. I understand that the user needs to know whether his complement tasks during this period will affect other tasks before selecting the maximum degree of parallelism.

File: dolphinscheduler-ui/src/views/projects/workflow/definition/components/start-modal.tsx
Comment (zhuangchong) [reply dolphinscheduler-ui/src/views/projects/workflow/definition/components/start-modal.tsx]: It is a very good solution for Master/Worker to have its own protection strategy, but this part needs to be strengthened. At the same time, 1500+ complement workflows were submitted, I saw that the Worker service was directly down.

Comment (ruanwenjun) [reply dolphinscheduler-ui/src/views/projects/workflow/definition/components/start-modal.tsx]: You can try to set `task-execute-threads-full-policy` to be REJECT, then the worker will only receive 100 tasks.
https://github.com/apache/dolphinscheduler/blob/d1135eabc7df9358fd210e423785d1067af64698/dolphinscheduler-worker/src/main/resources/application.yaml#L66-L66

BTW, I agree with @Gallardot if we want to set a maximal restriction, it's better to make this value to be configurable.

File: dolphinscheduler-ui/src/views/projects/workflow/definition/components/use-form.ts
Comment (ruanwenjun) [reply dolphinscheduler-ui/src/views/projects/workflow/definition/components/use-form.ts:69]: OK, this make sense to me, the maximum paramllelism does really will affect the current running scheduler workflow. We should direct users to try not to use such large parallel reads unless there is some special reason.

File: dolphinscheduler-ui/src/views/projects/workflow/definition/components/start-modal.tsx
Comment (SbloodyS) [reply dolphinscheduler-ui/src/views/projects/workflow/definition/components/start-modal.tsx]: > You can try to set `task-execute-threads-full-policy` to be REJECT, then the worker will only receive 100 tasks.
> 
> https://github.com/apache/dolphinscheduler/blob/d1135eabc7df9358fd210e423785d1067af64698/dolphinscheduler-worker/src/main/resources/application.yaml#L66-L66
> 
> BTW, I agree with @Gallardot if we want to set a maximal restriction, it's better to make this value to be configurable.

+1

Comment (calvinjiang) [reply dolphinscheduler-ui/src/views/projects/workflow/definition/components/start-modal.tsx]: I'v improved the front-end parallelism strategy so that it won't be limited to a default max value.  It'll just offer a tip.
image

Comment (zhuangchong) [reply dolphinscheduler-ui/src/views/projects/workflow/definition/components/start-modal.tsx]: Because Master/Worker has its own protection strategy, giving user prompts is a solution.

**Review by zhuangchong — APPROVED** [review]
LGTM.

**Review by songjianet — APPROVED** [review]

File: dolphinscheduler-ui/src/views/projects/workflow/definition/components/start-modal.tsx
Comment (ruanwenjun) [reply dolphinscheduler-ui/src/views/projects/workflow/definition/components/start-modal.tsx]: The warning message is too strange, why it's better not to be over 10?

**Pull request merged**
